docs/advanced/cast/custom.rst

```diff
@@ -38,56 +56,82 @@ type is explicitly allowed.
 
 .. code-block:: cpp
 
-    namespace PYBIND11_NAMESPACE { namespace detail {
-        template <> struct type_caster<inty> {
-        public:
-            /**
-             * This macro establishes the name 'inty' in
-             * function signatures and declares a local variable
-             * 'value' of type inty
-             */
-            PYBIND11_TYPE_CASTER(inty, const_name("inty"));
-
-            /**
-             * Conversion part 1 (Python->C++): convert a PyObject into a inty
-             * instance or return false upon failure. The second argument
-             * indicates whether implicit conversions should be applied.
-             */
-            bool load(handle src, bool) {
-                /* Extract PyObject from handle */
-                PyObject *source = src.ptr();
-                /* Try converting into a Python integer value */
-                PyObject *tmp = PyNumber_Long(source);
-                if (!tmp)
+    namespace pybind11 {
+    namespace detail {
+
+    template <>
+    struct type_caster<user_space::Point2D> {
+        // This macro inserts a lot of boilerplate code and sets the default type hint to `tuple`
+        PYBIND11_TYPE_CASTER(user_space::Point2D, const_name("tuple"));
+        // `arg_name` and `return_name` may optionally be used to specify type hints separately for
+        // arguments and return values.
+        // The signature of our negate function would then look like:
+        // `negate(Sequence[float]) -> tuple[float, float]`
+        static constexpr auto arg_name = const_name("Sequence[float]");
```

Comment:
`Sequence[int|float]` or `Sequence[Union[int, float]]`?

Reply:
According to [PEP 484 – Type Hints](https://peps.python.org/pep-0484/#the-numeric-tower) and the [typing documentation](https://typing.readthedocs.io/en/latest/spec/special-types.html#special-cases-for-float-and-complex):

> when an argument is annotated as having type float, an argument of type int is acceptable

So, if we want a Sequence of floats and/or ints, I think, `Sequence[float]` is correct and shorter.
But I do not mind if you prefer to include `int`, as it is still correct, but just adds a little redundancy.

Reply:
Although, now I see that I also used `Union[float,int]` in the `RealNumber` caster for the tests in [tests/test_pytypes.cpp](https://github.com/pybind/pybind11/pull/5450/files#diff-20bf1cca35b241c87fbf4ced7a3b55fd2421389f6861d06f67415c32d8c6bbc6).
However, I think that is ok there to just test the behavior of `arg_name`/`return_name`.
If you can think of a more sensible test class, I am open.

Reply:
> According to PEP 484 – Type Hints and the typing documentation:
>> when an argument is annotated as having type float, an argument of type int is acceptable

I wasn't aware of that, thanks for the explanation.

> However, I think that is ok there to just test the behavior of `arg_name`/`return_name`.

Could you maybe add a terse comment, that `int` isn't really needed, but appears for testing purposes? — For people who may use the test as a starting point for their work.

But looks good to me as is.

Reply:
I have added a comment

Reply:
Perfect, thanks!